radixdlt/src/integration/java/com/radixdlt/consensus/deterministic/MessageQueue.java

```diff
@@ -0,0 +1,180 @@
+
+/**
+ * Queue for messages by view.
+ */
+final class MessageQueue {
+	// Should really be an AbstractSequentialList, but I don't really want to have
+	// to implement a ListIterator.  Performance seems adequate for now.
+	final class ArrayOfLinkedLists<T> extends AbstractList<T> {
+
+		private final Object[] lists;
+		private final int size;
+
+		ArrayOfLinkedLists(Collection<LinkedList<T>> lists) {
+			this.lists = lists.toArray();
+			int count = 0;
+			for (int i = 0; i < this.lists.length; ++i) {
+				count += list(i).size();
+			}
+			this.size = count;
+		}
+
+		@SuppressWarnings("unchecked")
+		private LinkedList<T> list(int i) {
+			return (LinkedList<T>) lists[i];
+		}
+
+		@Override
+		public T get(int index) {
+			int currentIndex = index;
+			for (int i = 0; i < this.lists.length; ++i) {
+				LinkedList<T> list = list(i);
+				int listSize = list.size();
+				if (currentIndex < listSize) {
+					return list.get(currentIndex);
+				} else {
+					currentIndex -= listSize;
+				}
+			}
+			throw new ArrayIndexOutOfBoundsException(index);
+		}
+
+		@Override
+		public int size() {
+			return this.size;
+		}
+	}
+
+	private final HashMap<MessageRank, HashMap<ChannelId, LinkedList<ControlledMessage>>> messageQueue = Maps.newHashMap();
+	private MessageRank minimumMessageRank = null; // Cached minimum view
+
+	MessageQueue() {
+		// Nothing here for now
+	}
+
+	void add(MessageRank msgRank, ControlledMessage item) {
+		this.messageQueue.computeIfAbsent(msgRank, k -> Maps.newHashMap()).computeIfAbsent(item.getChannelId(), k -> Lists.newLinkedList()).add(item);
+		if (this.minimumMessageRank == null || msgRank.compareTo(this.minimumMessageRank) < 0) {
+			this.minimumMessageRank = msgRank;
+		}
+	}
+
+	ControlledMessage pop(ChannelId channelId) {
+		HashMap<ChannelId, LinkedList<ControlledMessage>> msgMap = this.messageQueue.get(this.minimumMessageRank);
+		if (msgMap == null) {
+			return painfulPop(channelId);
+		}
+		LinkedList<ControlledMessage> msgs = msgMap.get(channelId);
+		if (msgs == null) {
+			return painfulPop(channelId);
+		}
+		ControlledMessage item = msgs.pop();
+		if (msgs.isEmpty()) {
+			msgMap.remove(channelId);
+			if (msgMap.isEmpty()) {
+				this.messageQueue.remove(this.minimumMessageRank);
+				this.minimumMessageRank = minimumKey(this.messageQueue.keySet());
+			}
+		}
+		return item;
+	}
+
+	// Really only here to work with processNextMsg(int, int, Class<?>) in BFTDeterministicTest
+	private ControlledMessage painfulPop(ChannelId channelId) {
+		List<Map.Entry<MessageRank, HashMap<ChannelId, LinkedList<ControlledMessage>>>> entries = Lists.newArrayList(this.messageQueue.entrySet());
+		Collections.sort(entries, Map.Entry.comparingByKey());
+		for (Map.Entry<MessageRank, HashMap<ChannelId, LinkedList<ControlledMessage>>> entry : entries) {
+			HashMap<ChannelId, LinkedList<ControlledMessage>> msgMap = entry.getValue();
+			LinkedList<ControlledMessage> msgs = msgMap.get(channelId);
+			if (msgs != null) {
```

Comment:
Maybe we can avoid pyramid of doom as much as possible? This is easily done by using `continue`, like so:

Instead of:
```java
if (msgs != null) {
    // Oh no, I have to indent all my code one level!
}
```

We can do:

```java
if (msgs == null) {
    continue;
}
// Now code does not need to be indented :) 
```